src/mapblock_mesh.cpp

```diff
@@ -283,20 +319,62 @@ static u16 getSmoothLightCombined(const v3s16 &p, MeshMakeData *data)
 /*
 	Calculate smooth lighting at the given corner of p.
 	Both light banks.
+	Node at p is solid, and thus the lighting is face-dependent.
+*/
+u16 getSmoothLightSolid(const v3s16 &p, const v3s16 &face_dir, const v3s16 &corner, MeshMakeData *data)
+{
+	v3s16 neighbor_offset1, neighbor_offset2;
+
+	if (face_dir.X != 0) {
+		neighbor_offset1 = v3s16(0, corner.Y, 0);
+		neighbor_offset2 = v3s16(0, 0, corner.Z);
+	} else if (face_dir.Y != 0) {
+		neighbor_offset1 = v3s16(0, 0, corner.Z);
+		neighbor_offset2 = v3s16(corner.X, 0, 0);
+	} else if (face_dir.Z != 0) {
+		neighbor_offset1 = v3s16(corner.X,0,0);
+		neighbor_offset2 = v3s16(0,corner.Y,0);
+	}
+
+	const v3s16 dirs8[8] = {
+		// Always shine light
+		neighbor_offset1 + face_dir,
+		neighbor_offset2 + face_dir,
+		v3s16(0,0,0),
+		face_dir,
+
+		// Can be obstructed
+		neighbor_offset1 + neighbor_offset2 + face_dir,
+
+		// Only for ambient occlusion
+		neighbor_offset1,
+		neighbor_offset2,
+		neighbor_offset1 + neighbor_offset2
```

Comment:
Comments here describe the usage; but it should be easily understandable how are these nodes arranged.